### PR TITLE
Compiler refactor attach comments

### PR DESCRIPTION
**Pull request by petebacondarwin**
~~Blocked on top of #38866~~

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/translator/src/translator.ts [hunk 1]
```diff
@@ -387,6 +388,11 @@ export class ExpressionTranslatorVisitor<TStatement, TExpression> implements o.E
       T {
     return this.factory.setSourceMapRange(ast, createRange(span));
   }
+
+  private attachComments(statement: TStatement, leadingComments: LeadingComment[]): TStatement {
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/translator/src/translator.ts:392]: Can this be

```suggestion
  private attachComments(statement: TStatement, leadingComments: o.LeadingComment[]): TStatement {
```

and remove the (deep) import?

File: packages/compiler/src/output/output_ast.ts [hunk 2]
```diff
@@ -1003,10 +1003,18 @@ export class JSDocComment extends LeadingComment {
   }
 }
 
+/**
+ * A shared empty array that is used in Statements that do not have leading comments.
+ *
+ * This is frozen so that it cannot be mutated accidentally.
+ */
+const NO_LEADING_COMMENTS: LeadingComment[] = [];
```
Comment (JoostK) [inline packages/compiler/src/output/output_ast.ts:1011]: I don't quite see the value in this approach over the explicit `undefined` case we had before. The latter provides type-safety for the case where we haven't allocated an array yet, which won't be the case now.

Comment (petebacondarwin) [reply packages/compiler/src/output/output_ast.ts:1011]: What do you mean by type-safety here?

Comment (JoostK) [reply packages/compiler/src/output/output_ast.ts:1011]: You won't get a compile time error if you were to push to the frozen array, only a runtime failure. A null value would communicate this state in the type system.

Comment (petebacondarwin) [reply packages/compiler/src/output/output_ast.ts:1011]: Oh right. So I was considering whether to use a getter for the array and make it readonly, instead. WDYT?
I'll have a play with it tomorrow.

Alternatively we could bail on this commit altogether and just change the `translator.attachComments()` to accept `undefined` while allowing the `AstFactory` to require a defined array.

Comment (JoostK) [reply packages/compiler/src/output/output_ast.ts:1011]: I think I'd prefer dropping the commit and handling it in the translator.

Comment (petebacondarwin) [reply packages/compiler/src/output/output_ast.ts:1011]: Yeah, I agree.

**Review by JoostK — APPROVED** [review]
This is much cleaner indeed!

**Review by gkalpak — APPROVED** [review]
Yay! Much cleaner 🚀

Nit: There's a typo in the commit message (`it [...] return a statement` --> `it [...] returns a statement`)

**Pull request closed without merging**